### DISCUSSION TITLE
.git folder in homebrew-core takes too much storage space

### DISCUSSION THREAD

**Discussion by Mediocris** [Homebrew/discussions#2435] [Everyday usage] [answered]
It takes a long time to execute `brew update` command and I find that `.git` folder in `homebrew-core` takes too much space. Is there any way to reduce the file size?
<img width="791" alt="截屏2021-11-11 下午3 12 55" src="https://user-images.githubusercontent.com/38163902/141254284-2f928086-5a27-49ca-8965-d81ceffdc963.png">
<img width="791" alt="截屏2021-11-11 下午3 10 40" src="https://user-images.githubusercontent.com/38163902/141254127-7cd88e6a-c372-47ab-8f56-f04a7609352a.png">

**Comment by carlocab** [comment, marked as answer]
If you don't contribute to Homebrew/core, you can use `HOMEBREW_INSTALL_FROM_API`. This would allow you to `brew untap homebrew/core`. See `man brew` for details.

**Reply by Mediocris** [reply]
Thank you very much for your quick response, I have set the `HOMEBREW_INSTALL_FROM_API` and excute `brew untap homebrew/core`.
`brew info` dosen't work after that and `brew doctor` returned some warnings. How can I fix the problems?
<img width="885" alt="截屏2021-11-11 下午6 30 00" src="https://user-images.githubusercontent.com/38163902/141282914-bf00528a-cad7-460c-b854-6f9db00637e3.png" width="50%" height="50%">
<img width="728" alt="截屏2021-11-11 下午6 33 45" src="https://user-images.githubusercontent.com/38163902/141283220-7b2968e6-ef66-4d95-87f3-fcf887279a11.png" width="50%" height="50%">

**Reply by SMillerDev** [reply]
Try doing `brew update`

Tagging in @Rylan12 for more ideas

**Reply by carlocab** [reply]
`brew info` not working when a formula isn't installed is a known issue: https://github.com/Homebrew/brew/issues/12357

**Reply by Rylan12** [reply]
@Mediocris can you provide the output of `brew config`?

This shouldn't be happening (it was fixed in one of the recent versions of Homebrew) and I can't reproduce it on my machine.

**Reply by Mediocris** [reply]
> @Mediocris can you provide the output of `brew config`?
> 
> This shouldn't be happening (it was fixed in one of the recent versions of Homebrew) and I can't reproduce it on my machine.

@Rylan12 I've already updated `brew` to the latest version.
`brew info` still not working.

```
HOMEBREW_VERSION: 3.3.3
ORIGIN: https://github.com/Homebrew/brew.git
HEAD: 64fb3c86f87eaf4a4b2ba64ddec0a5a3f2530b4c
Last commit: 4 days ago
Core tap: N/A
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CASK_OPTS: []
HOMEBREW_CORE_GIT_REMOTE: https://github.com/Homebrew/homebrew-core
HOMEBREW_INSTALL_FROM_API: set
HOMEBREW_MAKE_JOBS: 12
Homebrew Ruby: 2.6.8 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: dodeca-core 64-bit kabylake
Clang: 13.0.0 build 1300
Git: 2.30.1 => /Library/Developer/CommandLineTools/usr/bin/git
Curl: 7.77.0 => /usr/bin/curl
macOS: 12.0.1-x86_64
CLT: 13.1.0.0.1.1633545042
Xcode: N/A 
```

**Reply by Rylan12** [reply]
Do you still have the output of `brew untap homebrew/core`? Can you run it again and post the output?

It looks like `homebrew/core` wasn't removed for some reason.

**Reply by Mediocris** [reply]
> Do you still have the output of `brew untap homebrew/core`? Can you run it again and post the output?
> 
> It looks like `homebrew/core` wasn't removed for some reason.

Here's the output of `brew untap homebrew/core`.
```
$ brew untap homebrew/core
  Error: No available tap homebrew/core.
```

**Reply by Rylan12** [reply]
Huh. Can you post the output of `ls -la /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core` and `git -C /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core status`?

I'd say just go ahead and manually remove the tap by running `rm -rf /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core`. Hopefully that should solve the problem

**Reply by Mediocris** [reply]
I just ran the commands you provided. I think the `homebrew-core` has been completely removed.

Sadly, the problem still exists. 😭😭😭

```
$ ls -la /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core
lsd: /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core: No such file or directory (os error 2).

$ git -C /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core status
fatal: cannot change to '/usr/local/Homebrew/Library/Taps/homebrew/homebrew-core': No such file or directory

$ rm -rf /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core

$ brew update
Already up-to-date.

$ brew info tmux
Error: No available formula or cask with the name "tmux".
```

**Reply by Rylan12** [reply]
Ah, I think we've had a miscommunication. As @carlocab mentioned [above](https://github.com/Homebrew/discussions/discussions/2435#discussioncomment-1625190), `brew info` not working is a known issue.

I was more concerned by the `brew doctor` failures. Are those gone?

**Reply by Mediocris** [reply]
`brew doctor` works fine. 

It returns some warnings about formulaes I have installed. Should I ignore those warnings?

```
$ brew doctor

Please note that these warnings are just used to help the Homebrew maintainers 
with debugging if you file an issue. If everything you use Homebrew for is 
working fine: please don't worry or file an issue; just ignore this. Thanks! 

Warning: Some installed kegs have no formulae!
This means they were either deleted or installed with 'brew diy'
You should find replacements for the following formulae:
 pyenv
 pkg-config
 pandoc
 cmake
 readline
 sqlite
 lsd
 zlib
 xz
 ca-certificates
 autoconf
 m4
 openssl@1.1
 pyenv-virtualenv
```

**Reply by Rylan12** [reply]
Yes, those should be fine to ignore unless you're having problems (other than the `brew info` issue). Do you know if those warnings showed up in the `brew doctor` output _before_ you set `HOMEBREW_INSTALL_FROM_API`? Do you know whether there are any formulae you have installed the _aren't_ on that list? You can compare the output of `brew list --formula` to check.

The reason I ask is that I thought I had fixed `brew doctor` so warnings wouldn't be shown when removing `homebrew/core` so if you were seeing the issues before then they're likely unrelated.

Can you also post the output to `cat $(brew --prefix)/opt/openssl@1.1/INSTALL_RECEIPT.json`?

**Reply by Mediocris** [reply]
> Do you know if those warnings showed up in the `brew doctor` output _before_ you set `HOMEBREW_INSTALL_FROM_API`?

No, I have `homebrew/core` on my computer before I set `HOMEBREW_INSTALL_FROM_API`. The warning information just shows up in a short time after I untapped `homebrew/core` and `brew doctor` works fine now. I think _doctor warning_ issue has been properly resolved.

> Do you know whether there are any formulae you have installed the _aren't_ on that list?

No, all formulaes I installed have been listed in the warning information.

> Can you also post the output to cat $(brew --prefix)/opt/openssl@1.1/INSTALL_RECEIPT.json?

```
$ cat $(brew --prefix)/opt/openssl@1.1/INSTALL_RECEIPT.json

{
  "homebrew_version": "3.2.17-96-g8fa2d99",
  "used_options": [

  ],
  "unused_options": [

  ],
  "built_as_bottle": true,
  "poured_from_bottle": true,
  "installed_as_dependency": true,
  "installed_on_request": false,
  "changed_files": [
    "bin/c_rehash",
    "CHANGES",
    "lib/pkgconfig/libcrypto.pc",
    "lib/pkgconfig/libssl.pc",
    "lib/pkgconfig/openssl.pc",
    "share/man/man1/ca.1ssl"
  ],
  "time": 1635598595,
  "source_modified_time": 1629812327,
  "compiler": "clang",
  "aliases": [

  ],
  "runtime_dependencies": [
    {
      "full_name": "ca-certificates",
      "version": "2021-10-26",
      "declared_directly": true
    }
  ],
  "source": {
    "spec": "stable",
    "versions": {
      "stable": "1.1.1l",
      "head": null,
      "version_scheme": 1
    },
    "path": "/usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/openssl@1.1.rb",
    "tap_git_head": "229487f07e126aa79b53d56ba36339ab88bafcbc",
    "tap": "homebrew/core"
  },
  "arch": "x86_64",
  "built_on": {
    "os": "Macintosh",
    "os_version": "macOS 12",
    "cpu_family": "penryn",
    "xcode": "13.1",
    "clt": "13.1.0.0.1.1633545042",
    "preferred_perl": "5.30"
  }
}%
```

**Reply by Rylan12** [reply]
Great, glad to hear it!

**Reply by n-osennij** [reply]
> If you don't contribute to Homebrew/core, you can use `HOMEBREW_INSTALL_FROM_API`. This would allow you to `brew untap homebrew/core`. See `man brew` for details.

hi, how can I change congig to add HOMEBREW_INSTALL_FROM_API?

**Comment by Saklad5** [comment]
Having stumbled upon this while trying to find something else, I have an alternative recommendation:

```sh
git -C "$(brew --repository)" gc --aggressive
```

That will instruct Git to [optimize the hell][1] out of Homebrew's Git repository. Afterwards[^1], it will be both smaller and faster, and it will *stay* that way[^2].

[1]: https://git-scm.com/docs/git-gc#Documentation/git-gc.txt---aggressive

---

The reason this isn't done automatically, unless you configure it to of course, is that it will take longer than your average Git command. Expect this to take anywhere from seconds to hours, depending on your hardware and how much room for improvement there is.

[^1]: Assuming you didn't do something weird like disabling compression in your Git configuration files, of course. You'd presumably know if you did.

[^2]: This basically takes everything currently in the repository and reorganizes it to be as efficient as possible. Think of it like taking a bunch of distinct chunks of data and compacting them into one. Future data will add more chunks, but the existing data will *stay* optimized. As a bonus, running the command again will almost certainly take less time, since it has less to do.

**Comment by ghost** [comment]
How about providing a official 'mirror' to homebrew-core, which only keeps current formulas, discarding the 10+ years history?
Then compact the history of the mirror repo every year.

**Reply by SMillerDev** [reply]
That would defeat the purpose of using git in the first place. But you're welcome to make a mirror like this for yourself.

**Reply by Saklad5** [reply]
Just use a blobless or shallow clone.

**Reply by SMillerDev** [reply]
Please don't, GitHub explicitly requested Homebrew to stop doing that. You're better off experimenting with the new `HOMEBREW_INSTALL_FROM_API` variable: https://docs.brew.sh/Manpage#environment

**Reply by Saklad5** [reply]
You could do it locally: make a full clone of Homebrew from GitHub, then make a shallow local clone and delete the full clone.

**Reply by ghost** [reply]
I would make my own mirror to homebrew-core and shallow clone from that. Could this error be eliminated when `$HOMEBREW_CORE_GIT_REMOTE` is set to a non-public service?
```
% brew update
Error: 
  homebrew-core is a shallow clone.
To `brew update`, first run:
```
